### DISCUSSION TITLE
Use case: intracerebral haemorrhage

### DISCUSSION THREAD

**Discussion by paulwrightwork** [Project-MONAI/MONAILabel#640] [Show and tell]
I am a complete noob to MONAI in general, but an experienced brain imager. I will be making notes here on getting MONAIlabel working for segmentation of haemorrhagic lesions.

**Comment by paulwrightwork** [comment]
### Step 1: installing Slicer preview in Linux.
1. Download the preview .tar.gz from https://download.slicer.org/
2. Extract the tar: `tar xzf Slicer-4.xx.yy.z-yyyy-mm-dd-linux-amd64.tar.gz`
  a. [Optional] Put the executable somewhere safe.
  b. `sudo cp -r Slicer-4.xx.yy.z-yyyy-mm-dd-linux-amd64 /usr/local/`
  c. `sudo ln /usr/local/bin/Slicer /usr/local/Slicer-4.xx.yy.z-yyyy-mm-dd-linux-amd64/Slicer`
  d. I believe the stable release of Slicer uses your home directory for settings files, but it appears the preview version does not, so I had to make it readable (probably not ideal, but only I use this machine): `sudo chmod -R o+w /usr/local/bin/Slicer /usr/local/Slicer-4.xx.yy.z-yyyy-mm-dd-linux-amd64/`
3. I found Slicer would not run because it was missing a shared library from ITK 5.3. This is the preview, so I had to build it!
  a. Download ITK 5.3rc02 source code from https://itk.org/download/
  b. `tar xzf InsightToolkit-5.3rc02.tar.gz`
  c. `mkdir itk-install && cd itk-install`
  d. You need to set a configuration option to build shared libraries:
  e. `cmake -DBUILD SHARED LIBS:BOOL=on ../InsightToolkit-5.3rc02.tar.gz` (takes a few minutes to configure)
  f. `make` (takes quite a few more minutes to compile)
  g. At this point you can put the compiled files where you need them. For my case I only took the shared libraries:
  h. `sudo cp lib/lib* /usr/local/lib`
  i. Slicer should now run. Phew!

**Reply by lassoan** [reply]
Sorry, there was a regression on Linux packaging of Slicer, introduced a few days ago when upgrading to latest ITK. We'll fix it ASAP. You can track the status of the issue here: https://github.com/Slicer/Slicer/issues/6202

**Reply by paulwrightwork** [reply]
@lassoan Do you know if the thing with config / ini files is a known issue? With my normal installation of Slicer, I was able to stick it in /usr/local as root and it worked. The preview one complained it could write to ini files or install plugins, so I assume the paths were set to somewhere in its /usr/local directory rather than ~/.config/NA-MIC, which is I think where the stable version puts things.

**Reply by lassoan** [reply]
Slicer works as any other virtual Python environments. You can either add a shared, system level Slicer application instance that is shared between all users (but can only modified using sudo), or add a user-specific Slicer application instance that the user can modify. You just cannot mix things: a user cannot install a Slicer extension (which may need to install additional Python packages) into system Slicer without sudo.

**Reply by paulwrightwork** [reply]
Thanks. That's intriguing to me, as I didn't have this problem with my earlier installation of Slicer. This was created in /usr/local using sudo but stored plugins and preference files in ~/.config/NA-MIC so I could install plugins and change preferences as a user. It was a while ago I installed it, so I don't remember exactly the steps. If I manage to retrace them and find anything interesting I'll let you know. I know I didn't manually set it to use ~/.config/NA-MIC, because I wasn't even aware that directory existed until just now.

**Reply by lassoan** [reply]
We made Slicer a fully portable application/Python virtual environment maybe a year ago. Yes, there is a small price to pay - that you cannot have mixed environments (mixing _system_ Slicer with _user_ extensions and Python packages), but that did not work correctly anyway after pip-installing of Python packages became possible.

The current solution is very powerful: you have a fully self-contained, portable application, including all settings (even Slicer.ini - if you have one in the application folder), DICOM database, extensions, and Python packages. There is no need for installation, there is no need to activate the Python environment, etc., you just double-click on the Slicer executable and the application starts. You run Slicer from a USB stick or external drive, optionally with a preconfigured DICOM database, so Slicer can be used as a zero-footprint DICOM viewer.

**Comment by paulwrightwork** [comment]
### Step 2: installing MONAIlabel and initial setup
Just follow the friendly instructions [here](https://github.com/Project-MONAI/MONAILabel#installation)! In my case:
```
mkdir /data/monailabel
cd /data/monailabel
virtualenv venv
source venv/bin/activate
pip install --upgrade pip setuptools wheel
pip install monailabel-weekly
```
> Note: I switched to the weekly version

I will be segmenting both haemorrhagic lesions and calcification from my images, so I chose `deepedit_multilabel`:
```
monailabel apps --download --name deepedit_multilabel --output apps
mv apps/deepedit_multilabel apps/deepedit_ich
```

I then modify the labels by editing [these lines](https://github.com/Project-MONAI/MONAILabel/blob/6fb539c44e1b89418db5d6ee459abcece709c78a/sample-apps/deepedit_multilabel/main.py#L43-L52) of main.py to:
```
        self.label_names = {
            "haemorrhage": 1,
            "calcification": 2,
            "background": 0,
        }
```

I have a set of CT brain images with expert labels. These are in [BIDS](https://bids-specification.readthedocs.io/en/stable/) format, so I have to reorganize them. Fortunately there is only one base image per subject, so I can get these easily:
```
mkdir my_dataset
find /bids_path/rawdata/ -name "sub-*.nii" -exec cp '{}' /data/monailabel/my_dataset \;
```

Unfortunately, my labels are in separate, binary images, so I need to combine them. The easiest way to do this is with the [FSL](https://fsl.fmrib.ox.ac.uk/) command line tool [find_the_biggest](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FDT/UserGuide#find_the_biggest). I also have MATLAB code that would do the trick: [maxmap.m.txt](https://github.com/Project-MONAI/MONAILabel/files/8097868/maxmap.m.txt) (remove the .txt). If you are trying this, it's your choice whether it's more painful to install FSL or to write a MATLAB wrapper for your images.

Example output for a single subject:
```
$ find_the_biggest sub-001_ses-1_label-haemorrhage_mask.nii sub-001_ses-1_label-calcium_mask.nii my_labels
number of inputs 3
Indices
1 sub-001_ses-1_label-haemorrhage_mask.nii
2 sub-001_ses-1_label-calcium_mask.nii
```

My script `get_labels.sh`:
```
#!/bin/bash
haemorrhage_image=$1
calcium_image=${haemorrhage_image/label-haemorrhage/label-calcium}
label_image=$(basename $haemorrhage_image)
label_image=${label_image/_label-haemorrhage_mask/_ct}
label_image=/data/monailabel/my_dataset/labels/final/${label_image}
find_the_biggest $haemorrhage_image $calcium_image $label_image
```

Command to run:
```
$ find /path/to/data/derivatives/manual/ -name "*label-haemorrhage*.nii" -exec ./get_labels.sh '{}' \;
```
(Note that FSL saves as .nii.gz, so in my case I'll need to either `gunzip` the labels, or `gzip` the base images.)

So now I have my app customized with my labels, and a full set of images and labels in the monailabel directory. To test it out I'll need to remove some of the manual images and see how it does with those.

**Comment by paulwrightwork** [comment]
### Step 3: start the server
Tried first with plain:
```
monailabel start_server --app apps/deepedit_ich --studies my_dataset
```
It didn't seem to use the values of `conf` that are in my app script, so now I'm going to try:
```
monailabel start_server \
  --app apps/deepedit_ich \
  --studies my_dataset \
  --conf use_pretrained_model 0 \
```
The second command prevented downloading the pretrained model as expected. Unfortunately, model training is stalling at about 5% complete.

**Comment by paulwrightwork** [comment]
### Results - one label
Using deepedit_multilabel, I trained the model on 46 images with the default 50 epochs in under one hour, leaving 6 for manual labelling. 

The first result is using only background and haemorrhage labels. The initial accuracy was 71%. The segmentation is pretty good, although it is blocky. The base images are thick sliced, with 0.5 x 0.5 x 5.0 mm voxels.

![haemorrhage_example](https://user-images.githubusercontent.com/32168177/155517352-69f9a208-d1ea-4378-8ce8-2cea3cb73937.png)

The main edits I needed to make involved trimming the edge of the automatic label. I found using Deepedit landmarks slow going, so mostly used the paint tool, switching to the draw tool for fine structures.

More to come ...

**Comment by paulwrightwork** [comment]
### Results - two labels

Training time was similar to one label, but overall accuracy was only 42%. This is because it failed to learn to spot the calcium labels, perhaps because these are small. Here is an example with model above and expert (not me) manual labels below.

Final accuracy after submitting my edits of the six remaining images went down to 31%!

![2label_example](https://user-images.githubusercontent.com/32168177/155519752-93eabf1c-b7df-48ff-93c4-19fa0218e0fe.png)

![2label_manual](https://user-images.githubusercontent.com/32168177/155519767-1291ee3f-7ec2-4261-a794-2fcd25968e77.png)
